File: server/services/taskService.js
const db = require("../db/jsonStore");
const { TaskError } = require("../utils/errors");
const activityService = require("./activityService");

const PRIORITIES = ["Low", "Medium", "High", "Urgent"];
const UPDATABLE_FIELDS = ["title", "description", "notes", "priority", "estimated_hours"];

function createTask(fields, createdBy) {
  const { title, description, notes, priority, estimated_hours } = fields;
  if (!title || !title.trim()) throw new TaskError("Title is required");
  if (priority && !PRIORITIES.includes(priority)) {
    throw new TaskError(`Priority must be one of: ${PRIORITIES.join(", ")}`);
  }
  if (estimated_hours !== undefined && estimated_hours !== null && Number(estimated_hours) <= 0) {
    throw new TaskError("Estimated hours must be a positive number");
  }

  const now = new Date().toISOString();
  const task = db.table("tasks").insert({
    title: title.trim(),
    description: description || "",
    notes: notes || "",
    priority: priority || "Medium",
    estimated_hours: estimated_hours ? Number(estimated_hours) : null,
    created_by: createdBy,
    is_deleted: false,
    created_at: now,
    updated_at: now,
  });

  activityService.log(createdBy, "CREATE", "task", task.id, null, task);
  return task;
}

function getTask(id) {
  const task = db.table("tasks").find(id);
  if (!task || task.is_deleted) throw new TaskError("Task not found", 404);
  return task;
}

function updateTask(id, fields, actorId) {
  const existing = getTask(id);
  const patch = {};
  for (const key of UPDATABLE_FIELDS) {
    if (fields[key] !== undefined) patch[key] = fields[key];
  }
  if (patch.priority && !PRIORITIES.includes(patch.priority)) {
    throw new TaskError(`Priority must be one of: ${PRIORITIES.join(", ")}`);
  }
  if (patch.estimated_hours !== undefined && patch.estimated_hours !== null && Number(patch.estimated_hours) <= 0) {
    throw new TaskError("Estimated hours must be a positive number");
  }
  patch.updated_at = new Date().toISOString();

  const updated = db.table("tasks").update(id, patch);
  activityService.log(actorId, "UPDATE", "task", id, existing, updated);
  return updated;
}

function deleteTask(id, actorId) {
  const existing = getTask(id);
  db.table("tasks").update(id, { is_deleted: true, updated_at: new Date().toISOString() });
  db.table("task_assignments").removeWhere((a) => a.task_id === Number(id));
  activityService.log(actorId, "DELETE", "task", id, existing, null);
  return true;
}

function listTasks(filters = {}) {
  let tasks = db.table("tasks").filter((t) => !t.is_deleted);

  if (filters.priority) tasks = tasks.filter((t) => t.priority === filters.priority);
  if (filters.created_by) tasks = tasks.filter((t) => t.created_by === Number(filters.created_by));

  if (filters.status) {
    const matchingTaskIds = new Set(
      db
        .table("task_assignments")
        .filter((a) => a.status === filters.status)
        .map((a) => a.task_id)
    );
    tasks = tasks.filter((t) => matchingTaskIds.has(t.id));
  }

  return tasks
    .slice()
    .sort((a, b) => new Date(b.created_at) - new Date(a.created_at))
    .map((task) => {
      const assignments = db.table("task_assignments").filter((a) => a.task_id === task.id);
      return Object.assign({}, task, { assignment_count: assignments.length });
    });
}

module.exports = { createTask, getTask, updateTask, deleteTask, listTasks, PRIORITIES };

File: backend/task_management/services/task_service.py
from datetime import datetime

from ..extensions import db
from ..models import Task, TaskAssignment
from .activity_service import log as log_activity
from .errors import TaskError

PRIORITIES = ("Low", "Medium", "High", "Urgent")
UPDATABLE_FIELDS = ["title", "description", "notes", "priority", "estimated_hours"]


def create_task(fields, created_by):
    title = (fields.get("title") or "").strip()
    if not title:
        raise TaskError("Title is required")

    priority = fields.get("priority") or "Medium"
    if priority not in PRIORITIES:
        raise TaskError(f"Priority must be one of: {', '.join(PRIORITIES)}")

    estimated_hours = fields.get("estimated_hours")
    if estimated_hours not in (None, "") and float(estimated_hours) <= 0:
        raise TaskError("Estimated hours must be a positive number")

    task = Task(
        title=title,
        description=fields.get("description") or "",
        notes=fields.get("notes") or "",
        priority=priority,
        estimated_hours=estimated_hours or None,
        created_by=created_by,
    )
    db.session.add(task)
    db.session.commit()

    log_activity(created_by, "CREATE", "task", task.id, None, task.to_dict())
    return task


def get_task(task_id):
    task = db.session.get(Task, task_id)
    if not task or task.is_deleted:
        raise TaskError("Task not found", 404)
    return task


def update_task(task_id, fields, actor_id):
    task = get_task(task_id)
    before = task.to_dict()

    if "priority" in fields and fields["priority"] and fields["priority"] not in PRIORITIES:
        raise TaskError(f"Priority must be one of: {', '.join(PRIORITIES)}")
    if fields.get("estimated_hours") not in (None, "") and "estimated_hours" in fields:
        if float(fields["estimated_hours"]) <= 0:
            raise TaskError("Estimated hours must be a positive number")

    for key in UPDATABLE_FIELDS:
        if key in fields and fields[key] is not None:
            setattr(task, key, fields[key] if fields[key] != "" else None)
    task.updated_at = datetime.utcnow()

    db.session.commit()
    log_activity(actor_id, "UPDATE", "task", task.id, before, task.to_dict())
    return task


def delete_task(task_id, actor_id):
    task = get_task(task_id)
    before = task.to_dict()

    task.is_deleted = True
    task.updated_at = datetime.utcnow()
    TaskAssignment.query.filter_by(task_id=task_id).delete()
    db.session.commit()

    log_activity(actor_id, "DELETE", "task", task_id, before, None)
    return True


def list_tasks(filters=None, page=1, per_page=100):
    """Returns `(tasks, total)`, each task already carrying its assignment count.

    The counts come from a single GROUP BY rather than one COUNT per row —
    the previous per-task query made the backlog screen O(n) round trips.
    """
    filters = filters or {}
    query = Task.query.filter_by(is_deleted=False)

    if filters.get("priority"):
        query = query.filter(Task.priority == filters["priority"])
    if filters.get("created_by"):
        query = query.filter(Task.created_by == filters["created_by"])
    if filters.get("search"):
        needle = f"%{filters['search'].strip()}%"
        query = query.filter(db.or_(Task.title.ilike(needle), Task.description.ilike(needle)))
    if filters.get("status"):
        query = query.filter(
            Task.id.in_(
                db.session.query(TaskAssignment.task_id).filter(
                    TaskAssignment.status == filters["status"]
                )
            )
        )

    total = query.order_by(None).count()
    page = max(1, page)
    tasks = (
        query.order_by(Task.created_at.desc())
        .limit(per_page)
        .offset((page - 1) * per_page)
        .all()
    )

    counts = dict(
        db.session.query(TaskAssignment.task_id, db.func.count(TaskAssignment.id))
        .filter(TaskAssignment.task_id.in_([t.id for t in tasks] or [0]))
        .group_by(TaskAssignment.task_id)
        .all()
    )

    return [t.to_dict(assignment_count=counts.get(t.id, 0)) for t in tasks], total
